### PR TITLE
Dynamically mark tests as slow/veryslow during pytest collection phase

### PR DESCRIPTION
**Pull request by bjodah**
Instead of cluttering tests with markers such as `@slow` this PR would allow us the check in a file `.ci/durations.json` which may be updated intermittently to contain slow tests.

I've grouped tests into 2 slow categories: `slow` and `veryslow`.
Running all tests would require passing both flags: ~~`pytest --slow --veryslow`~~ (`pytest --veryquickcheck` & `pytest --quickcheck`)
~~Now running pytest without flags runs~~ ~80% of the tests in ~100 seconds on my machine (with `--veryquickcheck`).

One idea that comes to mind is to have CircleCI (#12219) run the fast tests, with say the newest python version, which would complete within a minute or two giving fast feedback on users PRs. The CircleCI script could then trigger a full TravisCI build only when passing using a webhook.

There might be details here to be worked out before merging but I thought I this was an appropriate time for feedback, what do you think?

### CODE REVIEW

**Comment by asmeurer** [comment]
I'm not following the idea here. Is that file committed to the repo or not?

**Comment by bjodah** [comment]
It is committed (`.ci/durations.json`). We could of course host it on some server and have a hook to on the CI server to pull before tests and to update it after tests pass, but that's overengineering in my book.
The file isn't huge (100k) and compressess well under git (plain text json) and we only need to update it when we add slow tests and grow tired of running them among the quick tests. So I imagine this file being updated quite seldom (maybe every 100th or 1000th commit to the sympy repo).

**Comment by asmeurer** [comment]
I'm weary here. Every test that doesn't get run is a test that starts to fail eventually. Then we hit issues where slow tests have been failing for a while and it makes things like releases even slower than they already are.

**Comment by asmeurer** [comment]
Since it's json, why not use a dictionary that maps file names to a list of tests.

File: .ci/parse_durations_log.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+def main(limits=(10, .1)):
+    """
+    parses durations.log (see generate_durations_log.sh for how to generate that file)
+    """
+    groupings = [[] for _ in range(len(limits))]
+    for test_id, time in read_log():
+        for idx, lim in enumerate(limits):
+            if time >= lim:
+                groupings[idx].append(test_id)
+    open(os.path.join(ci_folder, 'durations.json'), 'wt').write(json.dumps([sorted(gr) for gr in groupings]))
```
Comment (asmeurer) [inline .ci/parse_durations_log.py]: Use a pretty sorted dump so that diffs are easier to read.

**Comment by bjodah** [comment]
We could use a dictionary to map to each file. It would make the plaintext file smaller (and maybe also a bit smaller after compression).
I've added a small print statement to show how long each group of tests take:
```
$ ./parse_durations_log.py 
accumulated_times: [13413.169999999998, 14896.230000000023]
```
so running with `--slow` takes 4.1 hours and `--veryslow` 3.7 hours on my machine.
I could tweak the limits so that `veryslow` runs considerably longer in total compared to `slow`.

My point is not that we should stop running slow tests, but it's nice to have an early exit when possible.
So a passing result on the fast tests should never be sufficient on the CI server, but failures among the fast tests allows for much quicker feedback.

What I'm striving for is essentially being able to run tests sorted by runtime, starting with the fastest and finishing with the slowest. But having a sorted list containing all tests will result in a quite big file (with lots of noise when updating).

**Comment by bjodah** [comment]
Changed to use a dictionary.
```
$ ./parse_durations_log.py 
number in group, accumulated_time: [(103, 13413.169999999998), (1366, 1483.060000000007)]
$ du -h durations.json 
76K	durations.json
```
So now there is an order of magnitude difference in running times between each of the options `pytest`, `pytest --slow` and `pytest --slow --veryslow`.
The json file is about 2000 lines long and easy to read. Using a dict made it a lot better.

**Comment by siefkenj** [comment]
As I wait patiently for Travis, this is sounding like a great idea!  A similar idea to what was stated earlier, you could have a `--quickceck` flag that specifies you only want the non-slow checks on CircleCI.  It's not an issue now, but you might also consider a whitelist.  For instance, `code_quality` should always be run.

**Comment by bjodah** [comment]
I made the skipping opt-in by using @siefkenj's suggested `--quickcheck` and also a `--veryquickcheck`.

I implemented a whitelist for pep8 and pyflakes tests (we could switch to a blacklist eventually - that would automatically require new modules to conform to pep8 and pyflakes).

I set the pep8 line width to 119 since this is what github allows to show before it starts scrolling (and it's a personal preference), but this can of course be adjusted.

The CircleCI build finishes in about 7 minutes after having run about 60% of the tests:
https://circleci.com/gh/bjodah/sympy/9

I'm having trouble getting `pytest.raises` to actually catch `SymPyDeprecationWarning`. I've tried a few things but haven't been able to make it work.

If we want to run all tests on CircleCI it's as easy as adding one more line (without the `--[very]quickcheck` flags).

**Comment by asmeurer** [comment]
I think he meant the tests in sympy/utilities/tests/test_code_quality.py. Using pep8 and/or pyflakes as part of our test suite is a completely separate discussion.

**Comment by bjodah** [comment]
I see, I misinterpreted it. Let's discuss that separately at another time then.
Right now pep8 and pyflakes are only enabled for a few files, I could reduce that to only include e.g. `conftest.py` if we believe the logic is worth having around, or I can strip it out, what do you think?

@asmeurer what do you think about the idea in general now that skipping is opt-in?

**Comment by asmeurer** [comment]
opt-in skipping is an interesting idea, although I'm unsure how much there will be a correlation between quickcheck passing and all tests passing.

**Comment by siefkenj** [comment]
@bjodah If you felt like doing some legwork, you could checkout several PRs where Travis has failed and see if quickcheck would catch those failures.  Or, perhaps you can just look at the tests that failed for them and see if they're on the quick list.

**Comment by bjodah** [comment]
@siefkenj, I've now looked at the most recent open pull requests marked with failed tests which
failed (but not those that timed out or due to theanocode printing error which seem
to have surfaced recently) to see if it would have helped:

#12378: if we include --doctest-modules
#12370: if we include --doctest-modules (2 failed commits)
#12354: yes (7 failed commits)
#12351: once I whitelist test_code_quality.py from durations.json (3 failed commits)
#12320: yes (10 failed commits)
#12296: if we run with python 2.7 as well
#12272: yes
#12231: yes (2 failed commits)
#12224: yes (5 failed commits)

Then I looked at the last 10 merged PRs:
#12393: (no failures)
#12384: (no failures)
#12360: if we run doctests and build docs (2 failed commits out of 4 in total)
#12347: if we run doctests
#12343: (no failures)
#12337: (no failures)
#12333: (no failures)
#12329: (no failures)
#12314: (no failures)

Then I looked at the last 10 closed PRs:
#12392: yes (1 failed commit)
#12382: once I whitelist test_code_quality.py
#12374: (canceled)
#12373: once I whitelist test_code_quality.py
#12367: (no failures)
#12364: once I whitelist test_code_quality.py 
#12357: yes (1 commit) and no (1 commit), latter would need white-listing of sympy/core/tests/test_args.py)
#12355: if we run doctests
#12353: (timed out)
#12350: yes (1 commit out of 2 tested)

(opening Travis logs on my laptop brings it to its knees...)

Another idea for even earlier feedback: do a git diff wrt to master and for each file changed:
1. if it matches `test_*.py` run tests on it.
2 "else if" it matches `*.py` run doctests on it.

**Comment by siefkenj** [comment]
@bjodah If I interpret your table correctly, we could cut the Travis builds by 60% if we use `--quickcheck` first and only run Travis if it succeeds?

**Comment by bjodah** [comment]
Wouldn't dare to give any exact number since it also depends on when the travis tests fail. The worst kind of failures are those that happen late during the CI build(s). When someone messes up big time the runtime are actually quite short (all tests exit early due to failed imports etc.).

Right now the backlog on travis is quite long, which means that most time is spent waiting for the tests to start. So if we expect this to make the load on the Travis tests low enough so that once they are sent there they start immediately - then it could give even greater savings.

And for negative tests the change may be even more dramatic: now a contributor might wait half a day before tests start, only to see them fail since they didn't know to e.g. run `test_code_quality.py`. With this setup they could get feedback within minutes.

**Comment by asmeurer** [comment]
This pull request is doing a lot of things at once. I think it would be easier to split it up into separate pull requests:

- Adding CircleCI support (https://github.com/sympy/sympy/issues/12219). This is probably a good testing ground for Circle, but I think we should consider doing all our tests on it, if it works well. As far as I know, it should be faster than Travis.

- Adding a quickcheck tests feature.

- Adding pyflakes/pep8 testing to the tests (which I disagree with for pep8).

- Moving CI testing to pytest.

**Comment by bjodah** [comment]
Yes, I know, it became a slippery slope. I've now changed this PR so that it only addresses the quick-check feature.

**Comment by asmeurer** [comment]
The changes now don't really affect anything else, so they look fine to me.

**Comment by smichr** [comment]
The last word from @asmeurer was that this looked fine. I would like to see things move in the direction of lightening the load on the service that we use for testing. So I'll commit what's here.

**Pull request merged**